File: apps/infra/providers/models/etherscan.py
"""
Etherscan API Data Models

Simple Pydantic models for Etherscan API responses focused on type safety.
"""

from typing import Any, Dict, List, Optional, Union
from pydantic import BaseModel, Field, ConfigDict
from enum import Enum


class EtherscanStatus(str, Enum):
    """Etherscan API response status codes."""

    SUCCESS = "1"
    FAILURE = "0"


class EtherscanTransaction(BaseModel):
    """Etherscan transaction model."""

    model_config = ConfigDict(populate_by_name=True)

    blockNumber: str
    timeStamp: str
    hash: str
    nonce: str
    blockHash: str
    transactionIndex: str
    from_address: str = Field(alias="from")
    to: Optional[str] = None
    value: str
    gas: str
    gasPrice: str
    isError: str
    txreceipt_status: str
    input: str
    contractAddress: Optional[str] = None
    cumulativeGasUsed: str
    gasUsed: str
    confirmations: str
    methodId: Optional[str] = None
    functionName: Optional[str] = None


class EtherscanProxyResponse(BaseModel):
    """Etherscan proxy API response model."""

    jsonrpc: str
    id: int
    result: Union[str, int, Dict[str, Any], List[Any], None]
    error: Optional[Dict[str, Any]] = None


class EtherscanLog(BaseModel):
    """Etherscan event log model."""

    address: str
    topics: List[str]
    data: str
    blockNumber: str
    timeStamp: str
    gasPrice: str
    gasUsed: str
    logIndex: str
    transactionHash: str
    transactionIndex: str


class EtherscanApiResponse(BaseModel):
    """Generic Etherscan API response model."""

    status: EtherscanStatus
    message: str
    result: Union[
        List[EtherscanTransaction],
        List[EtherscanLog],
        EtherscanTransaction,
        str,
        Dict[str, Any],
        None,
    ]

File: apps/dashboard/types/etherscan.ts
/**
 * Etherscan API Types
 * Based on https://docs.etherscan.io/api-endpoints/accounts
 */

// Base API Response Structure
export interface EtherscanApiResponse<T = any> {
  status: string;
  message: string;
  result: T;
}

// Common Transaction Fields
export interface EtherscanTransaction {
  blockNumber: string;
  timeStamp: string;
  hash: string;
  nonce: string;
  blockHash: string;
  transactionIndex: string;
  from: string;
  to: string;
  value: string;
  gas: string;
  gasPrice: string;
  gasUsed: string;
  cumulativeGasUsed: string;
  input: string;
  contractAddress: string;
  confirmations: string;
}

// Normal Transactions Response
export interface EtherscanNormalTransactionsResponse {
  status: string;
  message: string;
  result: EtherscanTransaction[];
}

// Token Transfer Events (ERC20, ERC721, ERC1155)
export interface EtherscanTokenTransfer {
  blockNumber: string;
  timeStamp: string;
  hash: string;
  nonce: string;
  blockHash: string;
  transactionIndex: string;
  gas: string;
  gasPrice: string;
  gasUsed: string;
  cumulativeGasUsed: string;
  input: string;
  contractAddress: string;
  from: string;
  to: string;
  tokenID?: string; // For ERC721/ERC1155
  tokenName: string;
  tokenSymbol: string;
  tokenValue?: string; // For ERC20/ERC1155
  tokenDecimal?: string; // For ERC20
  confirmations: string;
}

// Token Transfer Response
export interface EtherscanTokenTransfersResponse {
  status: string;
  message: string;
  result: EtherscanTokenTransfer[];
}

// Address Funded By Response
export interface EtherscanFundingInfo {
  block: number;
  timeStamp: string;
  fundingAddress: string;
  fundingTxn: string;
  value: string;
}

export interface EtherscanFundingResponse {
  status: string;
  message: string;
  result: EtherscanFundingInfo;
}

// Request Parameters
export interface EtherscanTransactionParams {
  address: string;
  chainId: number;
  startblock?: number;
  endblock?: number;
  page?: number;
  offset?: number;
  sort?: 'asc' | 'desc';
}

export interface EtherscanTokenTransferParams {
  address: string;
  chainId: number;
  contractaddress?: string;
  startblock?: number;
  endblock?: number;
  page?: number;
  offset?: number;
  sort?: 'asc' | 'desc';
}

export interface EtherscanFundingParams {
  address: string;
  chainId: number;
}

// Service Configuration
export interface EtherscanConfig {
  apiKey: string;
  timeout?: number;
}

// Rate limiting and error handling
export interface EtherscanRateLimit {
  callsPerSecond: number;
  callsPerDay: number;
  remainingCalls: number;
}

export interface EtherscanError {
  status: string;
  message: string;
  result: string;
}
